File: ts/src/timestamp/SystemClock.ts
import { Clock } from "./Clock";


/**
 * SystemClock reads the current instant from the operating-system wall clock.
 */
export class SystemClock implements Clock {
    private static _instance: SystemClock;

    public static getInstance(): SystemClock {
        if (!SystemClock._instance) {
            SystemClock._instance = new SystemClock();
        }

        return SystemClock._instance;
    }

    private constructor() {}

    /**
     * @returns The current Unix epoch second count from the system wall clock.
     */
    public nowUnixSeconds(): bigint {
        return BigInt(Math.floor(Date.now() / 1000));
    }
}

File: go/internal/timestamp/system_clock.go
package timestamp

import "time"

// SystemClock reads the current instant from the operating-system wall clock.
type SystemClock struct{}

var sharedSystemClock *SystemClock = &SystemClock{}

// SharedSystemClock returns the shared SystemClock singleton.
func SharedSystemClock() *SystemClock {
	return sharedSystemClock
}

// NowUnixSeconds returns the current Unix epoch second count from the system wall clock.
func (systemClock *SystemClock) NowUnixSeconds() uint64 {
	return uint64(time.Now().Unix())
}
